junit-platform-jfr/junit-platform-jfr.gradle.kts

```diff
@@ -0,0 +1,23 @@
+plugins {
+	`java-library-conventions`
+}
+
+description = "JUnit Platform Flight Recorder Support"
+
+dependencies {
+	internal(platform(project(":dependencies")))
+
+	api(platform(project(":junit-bom")))
+	api("org.apiguardian:apiguardian-api")
+	api(project(":junit-platform-launcher"))
+}
+
+javaLibrary {
+	mainJavaVersion = JavaVersion.VERSION_11
+}
+
+tasks {
+	verifyOSGi {
+		enabled=false
```

Comment:
When enabled, it yields:

```
Resolution failed. Capabilities satisfying the following requirements could not be found:
    [<<INITIAL>>]
      ? osgi.identity: (osgi.identity=junit-platform-jfr)
          ? [junit-platform-jfr version=1.7.0.SNAPSHOT]
              ? osgi.wiring.package: (&(osgi.wiring.package=jdk.jfr))
```

Reply:
@rotty3000 @bjhargrave Do you know what's wrong here?

Reply:
It is detecting a reference to the package `jdk.jfr` in the artifact `junit-platform-jfr`. And no artifacts in the resolution check which exports the package `jdk.jfr`.

I will let @rotty3000 suggest a solution as he provided the initial verifyOSGi task.

Reply:
```suggestion
		enabled = false
```

Reply:
I'm taking a look today.

Reply:
This may be due to Bnd not having the latest EE model properties which includes the new `jdk.jfr` package. We are just fixing that now in Bnd.

Reply:
Sounds like a potential cause - as other modules defined here that also don't export any package are working w/o special configuration directives. I guess.

Reply:
We will soon have a Bnd 5.1.1-SNAPSHOT build available that can be used to test if our fix solves this. I will let you know when it is available.

Reply:
Great! Take your time and looking forward to test-drive that build.

I might merge this PR soon -- with verifyOSGi disabled -- so, testing will happen on a new branch.

Reply:
I've confirmed that the upcoming bnd 5.1.1 will address the issue here.

Reply:
Via this commit? https://github.com/bndtools/bnd/commit/7e295c95a80202ca245f51939ed631775d62ee83

Reply:
For 5.1.1, it would technically be https://github.com/bndtools/bnd/commit/95ab05ca689bc2a442e3ebc4cf1efdf521a16abb which is the above commit cherry-picked onto the next branch which is where 5.1.1 currently lives.

Reply:
@bjhargrave I got a notification that 5.1.1 has been released but it's not on Maven Central, yet. Why is that?

Reply:
I made the build last night and released the staging repo this morning (EDT). So it should now be available on Maven Central now.

Reply:
Already indexed via https://search.maven.org/artifact/biz.aQute.bnd/biz.aQute.bnd/5.1.1/jar

Reply:
Thanks, I just upgraded `master` to 5.1.1. @sormuras Could you please rebase this branch to verify that the issue is now resolved?

Reply:
Rebase done and force-pushed.

Verifying that it works with `verifyOSGi` enabled locally ... in progress.

Reply:
This still does not bode well

```text
> Task :junit-platform-jfr:verifyOSGi FAILED
Resolution failed. Capabilities satisfying the following requirements could not be found:
    [<<INITIAL>>]
      ? osgi.identity: (osgi.identity=junit-platform-jfr)
          ? [junit-platform-jfr version=1.7.0.SNAPSHOT]
              ? osgi.wiring.package: (osgi.wiring.package=jdk.jfr)

Execution failed for task ':junit-platform-jfr:verifyOSGi'.
> D:\dev\github\junit-team\junit5\junit-platform-jfr\build\verifyOSGiProperties.bndrun resolution exception
```

Reply:
So I've just tested and there is only one minor tweak required to get past this now :)

in `buildSrc/src/main/kotlin/osgi-conventions.gradle.kts`

please make this small change:
```
@@ -65,11 +65,13 @@ val osgiPropertiesFile = file("$buildDir/verifyOSGiProperties.bndrun")
 // task writes out the properties necessary for it to verify the OSGi
 // metadata.
 val osgiProperties by tasks.registering(WriteProperties::class) {
 	outputFile = osgiPropertiesFile
 	property("-standalone", true)
-	property("-runee", "JavaSE-${Versions.jvmTarget}")
+	project.extensions.getByType(JavaLibraryExtension::class.java).let { javaLibrary ->
+		property("-runee", "JavaSE-${javaLibrary.mainJavaVersion}")
+	}
 	property("-runrequires", "osgi.identity;filter:='(osgi.identity=${project.name})'")
 	property("-runsystempackages", "jdk.internal.misc,sun.misc")
 }
 
 val osgiVerification by configurations.creating {

```
then finally:
```
@@ -18,9 +18,6 @@ javaLibrary {
 
 tasks {
 	compileModule {
 		targetCompatibility = "11"
 	}
-	verifyOSGi {
-		enabled = false
-	}
 }

```
With this all tests should now pass.

Reply:
This issue is that the Java versioned used for the "verify" task should match the _intended target_ for that code. Since each project seems to be able to adjust it's base Java level, we need to make `verifyOSGi` aware of that as well :) This small change fixes that instead of using a global preset version.

Reply:
If you would like I can push this as a PR to master branch including the update to 5.1.1 and then you can rebase?

Reply:
@rotty3000 Thanks for looking into this and providing such detailed instructions! I think we can take it from here. 👍 

That makes total sense. I'm afraid the error message was not very helpful for a non-OSGi user such as me. Is there a way to make it more verbose in case we run into a similar issue in the future?

Reply:
You're right that the error message was not informative. Who said "Software is hard."
But to be very honest, there are very few projects _I've seen_ in the wild which actually build artifacts targeting different versions of Java in a single build and also building for OSGi, so most times a global setting is more than enough. I had to dig into gradle/kotlin DSL to even figure out how to get the project specific java version :D So what you have here is bleeding edge Java Library + OSGi build spec. But the bnd team is trying very hard to make the whole thing less obtuse and esoteric. Thanks for putting up with us :)

Reply:
Applied the "small change" as suggested, @rotty3000 -- and works locally. At least the OSGi verification check.

Now that the integration tests are started, they revealed another quirk: https://bugs.openjdk.java.net/browse/JDK-8232092?focusedCommentId=14352730&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14352730 -- but that's another story.

Thanks for your support, Raymond!